File: backend/app/routers/matches.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_session
from app.models import Match, MatchStatus
from app.schemas.match import MatchDecisionResponse, MatchRead
from app.services.items import get_matches_for_user
from app.services.matching import approve_match, reject_match
from app.services.notifications import notify_match_resolution
from app.utils.dependencies import get_current_user

router = APIRouter(prefix="/matches", tags=["matches"])


@router.get("/my_matches", response_model=list[MatchRead])
async def my_matches(
    session: AsyncSession = Depends(get_session),
    current_user=Depends(get_current_user),
) -> list[MatchRead]:
    matches = await get_matches_for_user(session, user=current_user)
    return [MatchRead.model_validate(match) for match in matches]


def _ensure_pending_match(match: Match | None) -> Match:
    if match is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Match not found")
    if match.match_status != MatchStatus.PENDING:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Match already resolved")
    return match


@router.post("/{match_id}/approve", response_model=MatchDecisionResponse)
async def approve_match_endpoint(
    match_id: UUID,
    session: AsyncSession = Depends(get_session),
    current_user=Depends(get_current_user),
) -> MatchDecisionResponse:
    statement = select(Match).where(Match.id == match_id)
    match = _ensure_pending_match((await session.execute(statement)).scalar_one_or_none())

    try:
        match, loser_contact, finder_contact = await approve_match(session, match=match, acting_user=current_user)
    except PermissionError:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized to approve this match") from None

    await notify_match_resolution(match=match, loser_contact=loser_contact, finder_contact=finder_contact)

    await session.refresh(match, attribute_names=["lost_item", "found_item"])
    return MatchDecisionResponse(
        match=MatchRead.model_validate(match),
        message="Match approved and items archived",
        contact_shared_with_loser=loser_contact,
        contact_shared_with_finder=finder_contact,
    )


@router.post("/{match_id}/reject", response_model=MatchRead)
async def reject_match_endpoint(
    match_id: UUID,
    session: AsyncSession = Depends(get_session),
    current_user=Depends(get_current_user),
) -> MatchRead:
    statement = select(Match).where(Match.id == match_id)
    match = _ensure_pending_match((await session.execute(statement)).scalar_one_or_none())

    try:
        match = await reject_match(session, match=match, acting_user=current_user)
    except PermissionError:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized to reject this match") from None

    await session.refresh(match, attribute_names=["lost_item", "found_item"])
    return MatchRead.model_validate(match)

File: frontend/src/lib/api/matches.ts
import { apiRequest } from './client';
import type { Item } from './items';

export interface Match {
	id: string;
	lost_item_id: string;
	found_item_id: string;
	loser_id: string;
	finder_id: string;
	confidence_score: number;
	match_status: 'PENDING' | 'APPROVED' | 'REJECTED';
	created_at: string;
	lost_item: Item;
	found_item: Item;
}

export interface ApproveMatchResponse {
	match: Match;
	message: string;
	contact_shared_with_loser: {
		name: string;
		email: string;
		contact_number: string;
	};
	contact_shared_with_finder: {
		name: string;
		email: string;
		contact_number: string;
	};
}

export async function getMyMatches(): Promise<Match[]> {
	return apiRequest<Match[]>('/matches/my_matches');
}

export async function approveMatch(matchId: string): Promise<ApproveMatchResponse> {
	return apiRequest<ApproveMatchResponse>(`/matches/${matchId}/approve`, {
		method: 'POST'
	});
}

export async function rejectMatch(matchId: string): Promise<Match> {
	return apiRequest<Match>(`/matches/${matchId}/reject`, {
		method: 'POST'
	});
}

